src/producer/Producer.ts

```
@@ -0,0 +1,84 @@
+
+  protected deliveryReportCb: (err: Error, report: DeliveryReport) => Promise<void>;
+  protected readyCb: (topic: ReadyTopic, metaData: ReadyMetadata) => Promise<void>;
+
+  constructor(protected kafkaClient: KafkaProducer, private serializer: ISchemaRegistry, logger: Logger) {
+    super(logger);
+    this.kafkaClient = kafkaClient;
+  }
+
+  public async produce(topic: string, records: BaseAvroRecord[]) {
+    const encodedRecords = await this.encodeRecords(topic, records);
+    await this.assignCallbacks(topic, encodedRecords);
+    await this.kafkaClient.connect();
+  }
+
+  public onReady = (callback: (topic: ReadyTopic, metaData: ReadyMetadata) => Promise<void>) => {
+    this.readyCb = callback;
+  };
+
+  private assignCallbacks = async (topic: string, encodedRecords: Buffer[]) => {
+    await this.kafkaClient.on(KafkaEvents.ready, async (readyTopic: ReadyTopic, metaData: ReadyMetadata) => {
+      await this.ready(topic, encodedRecords, readyTopic, metaData);
+    });
+
+    if (this.deliveryReportCb) {
+      this.kafkaClient.on(KafkaEvents.deliveryReport, async (err, report) => {
+        await this.deliveryReportCb(err, report);
+      });
+    }
+
+    this.assignCommonCallbacks();
+  };
+
+  protected ready = async (
+    topic: string,
+    encodedRecords: Buffer[],
+    readyTopic: ReadyTopic,
+    metadata: ReadyMetadata
+  ): Promise<void> => {
+    if (this.readyCb) {
+      await this.readyCb(readyTopic, metadata);
+    }
+    await this.produceRecords(encodedRecords, topic);
+    await this.kafkaClient.flush(Producer.DEFAULT_FLUSH_TIMEOUT, null);
+    this.kafkaClient.disconnect();
+  };
+
+  private encodeRecords = async (topic: string, records: BaseAvroRecord[]): Promise<Buffer[]> =>
+    await Promise.all(
+      records.map(async record => {
+        try {
+          return await this.serializer.encodeMessage(topic, record.schema(), record);
+        } catch (e) {
+          this.logger.error(`Unable to encode record: ${record}`);
```

Hmm we may actually want to throw this error and handle it at a higher level. I think this would fall into the category of "invalid" records, and I think we may want to write those somewhere so that they are not lost.